KAFKA-7893 Refactor ConsumerBounceTest to reuse functionality from BaseConsumerTest

stanislavkozlovski opened the pull request:
This PR should help address the flakiness in the ConsumerBounceTest#testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup test (https://issues.apache.org/jira/browse/KAFKA-7965). I tested this locally and have verified it significantly reduces flakiness - 25/25 tests now pass. Running the test 25 times in trunk, I'd get `18/25` passes.

It does so by reusing the less-flaky consumer integration testing functionality inside `BaseConsumerTest`. Most notably, the test now makes use of the `ConsumerAssignmentPoller` class  - each consumer now polls non-stop rather than the more batch-oriented polling we had in `ConsumerBounceTest#waitForRebalance()`.

stanislavkozlovski committed:
Refactor ConsumerBounceTest to inherit BaseConsumerTest

Inheritance is ConsumerBounceTest->BaseConsumerTest->BaseRequestTest->IntegrationTestHarness

Also refactored testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup to use consumer pollers from BaseConsumerTest

Refactor more tests to use ConsumerPollers

testConsumerReceivesFatalExceptionWhenGroupPassesMaxSize and testClose specifically

Refactor rest of ConsumerBounceTest tests

Remove unused imports

Make the code a bit cleaner

Remove unused KafkaConfig import

Address PR comments

stanislavkozlovski commented:
retest this please

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -223,6 +236,96 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
+                                                   topicsToSubscribe: List[String],
+                                                   subscriptions: Set[TopicPartition],
+                                                   group: String = group): (mutable.Buffer[KafkaConsumer[Array[Byte], Array[Byte]]], mutable.Buffer[ConsumerAssignmentPoller]) = {
+    assertTrue(consumerGroup.size + numOfConsumersToAdd <= subscriptions.size)
```
Not sure why this should be a requirement. I think `validateGroupAssignment` can still be satisfied as long as the subscription is non-empty.

stanislavkozlovski replied:
`isPartitionAssignmentValid(assignments, subscriptions)`
```
* Check whether partition assignment is valid
   * Assumes partition assignment is valid if
   * 1. Every consumer got assigned at least one partition
   * 2. Each partition is assigned to only one consumer
   * 3. Every partition is assigned to one of the consumers
```

It checks 
```
val allNonEmptyAssignments = assignments.forall(assignment => assignment.nonEmpty)
    if (!allNonEmptyAssignments) {
      // at least one consumer got empty assignment
      return false
    }
```

I don't think it will pass if that assertion doesn't itself

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -253,13 +356,17 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
   protected class ConsumerAssignmentPoller(consumer: Consumer[Array[Byte], Array[Byte]],
-                                           topicsToSubscribe: List[String]) extends ShutdownableThread("daemon-consumer-assignment", false)
+                                           topicsToSubscribe: List[String],
+                                           partitionsToAssign: Set[TopicPartition] = Set.empty[TopicPartition]) extends ShutdownableThread("daemon-consumer-assignment", false)
```
Seems like we should take either `topicsToSubscribe` or `partitionsToAssign`, but not both. Maybe we can have two separate constructors. Also, `subscribe` below should probably throw an illegal state error if `partitionsToAssign` is not empty.

stanislavkozlovski replied:
I added two new secondary constructors - can you check?

hachikuji commented on core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala:
```diff
@@ -29,25 +26,23 @@ import org.apache.kafka.common.errors.GroupMaxSizeReachedException
-  val topic = "topic"
-  val part = 0
-  val tp = new TopicPartition(topic, part)
+class ConsumerBounceTest extends BaseConsumerTest with Logging {
```
Extending `BaseConsumerTest` means we inherit its test cases (i.e. `testSimpleConsumption` and `testCoordinatorFailover`). I wonder if we should move these tests somewhere else.

stanislavkozlovski replied:
We have 6 test classes that make use of those test cases. Only `ConsumerBounceTest` is redundant. I don't think it hurts that much.

We could create a new class `BasicConsumerTests` which defines those tests if you think that's a good approach

hachikuji replied:
Yeah, it is not that big of a deal. Just it's a bit annoying to extend the build time even more because of tests we know are redundant. I think I slightly prefer your suggestion for `BasicConsumerTests`. Or maybe we can call the new class `BaseConsumerTest` and rename the existing one to `AbstractConsumerTest`?

The use of inheritance in these test cases has been a bit problematic in general. It would probably be better to split out any shared utility logic into a separate utility class. We don't have to do that here though.

stanislavkozlovski replied:
I like your thinking. Let's leave this as it is for the time being so we can merge this PR and reduce the flakiness of ConsumerBounceTest.

I've opened up a follow-up JIRA to track this - https://issues.apache.org/jira/browse/KAFKA-8176

hachikuji commented on core/src/test/scala/integration/kafka/api/ConsumerBounceTest.scala:
```diff
@@ -306,148 +299,83 @@ class ConsumerBounceTest extends BaseRequestTest with Logging {
+            if (!thrownException.isInstanceOf[GroupMaxSizeReachedException]) {
+              throw thrownException
+            }
+            if (kickedConsumerOut) {
```
Could probably get rid of this variable and just use `kickedOutConsumerIdx.foreach`.

hachikuji reviewed (commented):
Nice cleanup! Left a few comments.

stanislavkozlovski commented:
@hachikuji apologies for squashing all the commits into one - I had to rebase and did not want to deal fixing merge conflicts with each commit. I shouldn't have done that as it now makes your review not as straightforward

stanislavkozlovski committed:
Check proper partitionsToAssign collection before throwing ISE in ConsumerPoller

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
(I am also running the tests locally)

stanislavkozlovski committed:
Fix failing tests due to incorrect consumer group usage

stanislavkozlovski committed:
Generalize IntegrationTestHarness#generateConfigs

Adds a new protected method modifyConfigs, mainly to allow the inheriting BaseRequestTest to change the way it handles configs in a more succinct way and SaslPlaintextConsumerTest to override its wanted listeners

Also rename BaseRequestTest#numBrokers to serverCount. This consolidates the use of the same method in IntegrationTestHarness

stanislavkozlovski commented:
The `SaslPlainPlaintextConsumerTest` class had problems due to not initializing its config listeners properly. This is because the `BaseRequestTest` overrides `generateConfigs`.
I generalized the code a bit more to allow all classes to modify exactly what they want

stanislavkozlovski commented:
retest this please

stanislavkozlovski committed:
Call super().modifyProps in SaslPlainPlaintextConsumerTest

hachikuji commented:
retest this please

hachikuji commented:
retest this please

ijuma commented:
@stanislavkozlovski you should use `merge` instead of `rebase`.

ijuma commented on core/src/test/scala/unit/kafka/server/BaseRequestTest.scala:
```diff
@@ -32,22 +32,19 @@ import org.apache.kafka.common.requests.{AbstractRequest, AbstractRequestRespons
+  override def serverCount: Int = 3
 
   // If required, override properties by mutating the passed Properties object
   protected def propertyOverrides(properties: Properties) {}
```
I think this name doesn't quite work generally since `IntegrationTestHarness` exposes brokers and clients. So it should probably be renamed to make it clear that it's only for broker properties.

stanislavkozlovski replied:
That's a great catch, thanks

stanislavkozlovski committed:
Rename 'propertyOverrides' -> 'brokerPropertyOverrides'

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
retest this please

stanislavkozlovski commented:
Testing this locally, I see the `ConsumerBounceTest#testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup` gets stuck at the very last `sendRecords()` call.
Each producer seems to throw:
```
org.apache.kafka.common.errors.TimeoutException: Topic group-max-size-test not present in metadata after 60000 ms.
```

The issue was that we would not call `restartDeadBrokers()` after we had found a kicked-out consumer and were essentially running a reduced cluster.

stanislavkozlovski committed:
Add missing restartDeadBrokers() call

hachikuji commented:
retest this please

stanislavkozlovski commented:
Strange, I could have sworn this built and passes checkstyle/findbugs locally. Sorry for the noise. I will rebase with master as well now as there seem to be conflicting files

stanislavkozlovski committed:
Merge branch 'trunk' into KAFKA-7893-refactor-consumer-bounce-test

hachikuji commented on core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala:
```diff
@@ -54,7 +54,7 @@ import scala.collection.mutable.Buffer
 class AuthorizerIntegrationTest extends BaseRequestTest {
 
-  override def numBrokers: Int = 1
+  override def serverCount: Int = 1
```
nit: we're a little inconsistent with naming (broker vs server). Perhaps this should be `brokerCount` to go along with `brokerPropertyOverrides`?

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -30,37 +31,43 @@ import org.apache.kafka.clients.producer.KafkaProducer
+  val group = "my-test"
   val producerClientId = "ConsumerTestProducer"
   val consumerClientId = "ConsumerTestConsumer"
+  val GroupMaxSessionTimeoutMs = 30000L
```
nit: why is only this one capitalized?

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -130,6 +137,12 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
 
+
+  protected def createConsumerWithGroupId(groupId: String): KafkaConsumer[Array[Byte], Array[Byte]] = {
+    consumerConfig.setProperty(ConsumerConfig.GROUP_ID_CONFIG, groupId)
```
It's a little surprising to find that we mutate `consumerConfig` here. So if you called `createConsumerWithGroupId` first and then `createConsumer`, the first call would affect the second call. Maybe instead we could duplicate `consumerConfig` and override group id?

stanislavkozlovski replied:
Good call. I'll make use of the `configOverrides` var in `createConsumer()`

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -223,6 +236,100 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
+    * @param consumerPollers current consumer pollers
+    * @param topicsToSubscribe topics to which new consumers will subscribe to
+    * @param subscriptions set of all topic partitions
+  */
```
nit: misaligned

hachikuji commented on core/src/test/scala/integration/kafka/api/BaseConsumerTest.scala:
```diff
@@ -308,16 +433,19 @@ abstract class BaseConsumerTest extends IntegrationTestHarness {
   /**
    * Check whether partition assignment is valid
-   * Assumes partition assignment is valid iff
+   * Assumes partition assignment is valid if
```
The original was probably intentional, if pedantic.

hachikuji reviewed (commented):
Sorry for the delay. Left a few additional comments.

stanislavkozlovski committed:
Address PR comments
rename serverCount => brokerCount
fix var naming capitalization
fix javadoc indentation and typo

hachikuji commented:
retest this please

hachikuji reviewed (approved):
LGTM. Merging to trunk.
